Import staff patrol areas from RCT1 save files

muemart opened the pull request:
This adds the necessary code to import patrol areas for staff from RCT1 saves/scenarios (e.g. Diamond Heights). Previously, the patrol areas were ignored and the staff wasn't correctly "registered" either, which resulted in #5445.
There's probably a more clever way to convert between the two patrol area formats (they are both large bitfields, but with different dimensions), but I don't think it's not necessary to try and optimize it.

Gymnasiast commented on src/openrct2/rct1.h:
```diff
@@ -721,8 +722,8 @@ typedef struct rct1_s4 {
     uint16 game_counter_5;
-    uint8 patrol_areas[0x3C00];
-    uint8 unk_1F42AA[116];
+    uint8 patrol_areas[(RCT1_MAX_STAFF + 4) * 128];
```
Why `+4`? And what is the meaning of 128?

muemart replied:
For every staff type, there is an additional "meta-patrol-area", which is the union of all patrol areas for that type. E.g. there is one combined patrol area of all janitors, one of all mechanics, etc. There are four staff types, that's why +4.
Every patrol area needs to be able to cover the entire park, and you need 128 uint8s for this (that's 1024 bits, and one bit represents one of the 4x4 squares. Max map size is 128x128, which is 32x32=1024 4x4 squares)
Should this be documented somehow? The RCT2 part is basically the same, but with a bigger map size, and isn't really documented either.

duncanspumpkin replied:
Yes please document that as well. We are trying to get rid of as many magic numbers as possible

muemart commented:
Okay, I added some defines for the magic numbers and changed all usages I could find. While doing that I also took the opportunity to simplify the usages of gStaffPatrolAreas.

Gymnasiast commented on src/openrct2/peep/staff.c:
```diff
@@ -469,13 +468,13 @@ void game_command_set_staff_patrol(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32
             return;
         }
         rct_peep *peep = &sprite->peep;
-        sint32 patrolOffset = peep->staff_id * (64 * 64 / 8);
```
64 * 64 / 8 = 512, not 128.

muemart replied:
Everything is divided by 4 again, because we're indexing into uint32s, but the pointer arithmetic works in bytes (same for the other cases).

Gymnasiast commented on src/openrct2/peep/staff.c:
```diff
@@ -578,9 +577,9 @@ void staff_update_greyed_patrol_areas()
-        uint32 *addr = (uint32*)((uintptr_t)gStaffPatrolAreas + ((staff_type + STAFF_MAX_COUNT) * 512));
-        for (sint32 i = 0; i < 128; i++) {
-            addr[i] = 0;
+        sint32 staffPatrolOffset = (staff_type + STAFF_MAX_COUNT) * RCT12_PATROL_AREA_SIZE;
```
See above.

Gymnasiast commented on src/openrct2/peep/staff.c:
```diff
@@ -589,9 +588,9 @@ void staff_update_greyed_patrol_areas()
-                uint32 *addr2 = (uint32*)((uintptr_t)gStaffPatrolAreas + (peep->staff_id * 512));
-                for (sint32 i = 0; i < 128; i++) {
-                    addr[i] |= addr2[i];
+                sint32 peepPatrolOffset = peep->staff_id * RCT12_PATROL_AREA_SIZE;
```
Ditto

Gymnasiast commented on src/openrct2/rct1.h:
```diff
@@ -721,8 +723,8 @@ typedef struct rct1_s4 {
     uint16 game_counter_5;
-    uint8 patrol_areas[0x3C00];
-    uint8 unk_1F42AA[116];
+    uint8 patrol_areas[(RCT1_MAX_STAFF + STAFF_TYPE_COUNT) * RCT12_PATROL_AREA_SIZE];
```
Needs a separate RCT12_STAFF_TYPE_COUNT, in case OpenRCT2 ever supports more types of staff (either directly or through plug-ins).

Gymnasiast commented on src/openrct2/windows/Staff.cpp:
```diff
@@ -571,9 +571,8 @@ void window_staff_overview_dropdown(rct_window *w, rct_widgetindex widgetIndex,
     // Clear patrol
     if (dropdownIndex == 1) {
         rct_peep* peep = GET_PEEP(w->number);
-        uint32 *addr = (uint32*)((uintptr_t)gStaffPatrolAreas + (peep->staff_id * 512));
```
Again, 512 vs. 128.

Gymnasiast commented on src/openrct2/rct12.h:
```diff
@@ -25,6 +25,8 @@
 #define RCT12_MAX_STATIONS_PER_RIDE 4
 #define RCT12_MAX_PEEP_SPAWNS       2
 #define RCT12_MAX_PARK_ENTRANCES    4
+#define RCT12_PATROL_AREA_SIZE      128
```
Could you add a comment explaining why it is 128?

Gymnasiast commented:
You introduced constants for RCT2_MAX_STAFF and RCT12_PATROL_AREA_SIZE. This is good, but we only use prefixed constants in the parts directly interfacing with S4/S6.

You'll need to define MAX_STAFF and PATROL_AREA_SIZE in staff.h (with the same value as  RCT2_MAX_STAFF and RCT12_PATROL_AREA_SIZE), and replace all occurrences that are not directly related to S4/S6 reading/writing, so we can safely increase the staff limit later on.

muemart commented:
Done. I also noticed a small inconsistency in what the defines meant, see commit message for a bit more details.

Gymnasiast commented on src/openrct2/peep/staff.c:
```diff
@@ -47,8 +47,8 @@ const rct_string_id StaffCostumeNames[] = {
 
-uint32 gStaffPatrolAreas[204 * 128];
-uint8 gStaffModes[204];
+uint32 gStaffPatrolAreas[(STAFF_MAX_COUNT + STAFF_TYPE_COUNT) * STAFF_PATROL_AREA_SIZE];
```
Please add a comment explaining this multiplication.

Gymnasiast commented on src/openrct2/rct1/S4Importer.cpp:
```diff
@@ -1294,6 +1294,19 @@ class S4Importer final : public IParkImporter
             FixPeepNextInQueue(peep, spriteIndexMap);
         }
+
+        // The RCT2 structures are bigger than in RCT1, so set them to zero
```
```The RCT2/OpenRCT2 structures are bigger than in RCT1, so set them to zero```

muemart commented:
I added some more additional documentation where I thought it was appropriate. Hopefully it's clear enough so that others can also understand what's going on.

Gymnasiast commented:
Thank you very much for your PR. With all the changes, we also improved the documentation and future-proofness of the code base. 👍
